File: src/models/energy_sensor.py
import random


class EnergySensor:

    VOLTAGE_VARIATION_PERCENTAJE = 0.01
    CURRENT_VARIATION_PERCENTAJE = 0.05
    PRECISION = 3

    def __init__(self, ref_voltage, ref_current):
        self.ref_voltage = ref_voltage
        self.ref_current = ref_current

    def get_voltage(self):
        return self.__get_variation(self.ref_voltage, self.VOLTAGE_VARIATION_PERCENTAJE)

    def get_current(self):
        return self.__get_variation(self.ref_current, self.CURRENT_VARIATION_PERCENTAJE)

    def __get_variation(self, reference, variation_percentaje):
        variation = reference + (reference * random.uniform(-variation_percentaje, variation_percentaje))
        return round(variation, self.PRECISION)

File: js-device-simulator/js/models/energy-sensor.js
import { roundToDecimal, randomBetween } from '../utils/math.js';
import { MEASURES_DECIMAL_PRESICION } from '../config.js';

export class EnergySensor {
    _VOLTAGE_VARIATION_PERCENTAJE = 0.01
    _CURRENT_VARIATION_PERCENTAJE = 0.05
    _PRECISION = 3

    _refVoltage = null;
    _refCurrent = null;

    constructor(refVoltage, refCurrent) {
        this._refVoltage = refVoltage;
        this._refCurrent = refCurrent;
    }

    getVoltage() {
        return this._getNumberVariation(this._refVoltage, this._VOLTAGE_VARIATION_PERCENTAJE);
    }

    getCurrent() {
        return this._getNumberVariation(this._refCurrent, this._CURRENT_VARIATION_PERCENTAJE);
    }

    _getNumberVariation(referenceValue, variationPercentaje) {
        const variation = referenceValue + randomBetween(-variationPercentaje, variationPercentaje);
        return roundToDecimal(variation, MEASURES_DECIMAL_PRESICION);        
    }
}